python/ray/data/_internal/plan.py

```diff
@@ -549,12 +523,6 @@ def execute(
                 # recording unnecessary metrics for an empty plan execution.
                 blocks = self._in_blocks
                 stats = _get_initial_stats_from_plan(self)
-            elif self.is_read_only():
-                # If the Dataset is read-only, get the LazyBlockList without
-                # executing the plan by only fetching metadata available from
-                # the input Datasource or Reader without executing its ReadTasks.
-                blocks = get_legacy_lazy_block_list_read_only(self)
-                stats = _get_initial_stats_from_plan(self)
```

Comment:
we still need to keep this logic right? otherwise how do we get metadata without executing the read tasks?

Reply:
We need metadata for methods like `Dataset.count()`, `Dataset.schema()`, and `Dataset.input_files()`. Rather than getting metadata from a `LazyBlockList`, the implementation in this PR gets metadata from the logical operators (e.g., `LogicalOperator.count()`).

Reply:
Specifically, we need the `LazyBlockList` metadata for the following methods:
```
Dataset.__repr__
Dataset.count
Dataset.schema
Dataset.input_files
Dataest.size_bytes
```

After this PR, `ExecutionPlan.execute` always actually performs execution, and there isn't a separate code path for reading data with `LazyBlockList`.

Reply:
OK got it, makes sense. thanks for the clarification